File: src/ipc/export.ts
import { invoke } from "@tauri-apps/api/core";

export type ExportScope = {
  interviewIds: number[] | null;
  tagIds: number[] | null;
};

const toRawScope = (s: ExportScope) => ({
  interview_ids: s.interviewIds,
  tag_ids: s.tagIds,
});

export const exportCsv = (
  scope: ExportScope,
  destination: string,
): Promise<void> =>
  invoke("export_csv", { scope: toRawScope(scope), destination });

export const exportMarkdown = (
  scope: ExportScope,
  destination: string,
): Promise<void> =>
  invoke("export_markdown", { scope: toRawScope(scope), destination });

export const exportRefi = (
  scope: ExportScope,
  destination: string,
): Promise<void> =>
  invoke("export_refi", { scope: toRawScope(scope), destination });

export const exportStats = (
  scope: ExportScope,
  format: "csv" | "markdown",
  destination: string,
): Promise<void> =>
  invoke("export_stats", { scope: toRawScope(scope), format, destination });

export const exportCodebook = (
  format: "json" | "csv",
  destination: string,
): Promise<void> => invoke("export_codebook", { format, destination });

File: src-tauri/src/commands/export.rs
use crate::commands::util::project_conn;
use crate::error::AppResult;
use crate::export::{self, ExportScope};
use serde::Deserialize;
use std::fs::File;
use std::path::PathBuf;

#[derive(Debug, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum StatsFormat {
    Csv,
    Markdown,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum CodebookFormat {
    Json,
    Csv,
}

#[tauri::command]
pub async fn export_csv(
    app: tauri::AppHandle,
    scope: ExportScope,
    destination: String,
) -> AppResult<()> {
    let conn = project_conn(&app)?;
    let data = export::compose(&conn, &scope)?;
    let mut f = File::create(PathBuf::from(destination))?;
    crate::export::csv_export::write_csv(&data, &mut f)
}

#[tauri::command]
pub async fn export_markdown(
    app: tauri::AppHandle,
    scope: ExportScope,
    destination: String,
) -> AppResult<()> {
    let conn = project_conn(&app)?;
    let data = export::compose(&conn, &scope)?;
    let mut f = File::create(PathBuf::from(destination))?;
    crate::export::markdown::write_markdown(&data, &mut f)
}

#[tauri::command]
pub async fn export_refi(
    app: tauri::AppHandle,
    scope: ExportScope,
    destination: String,
) -> AppResult<()> {
    let conn = project_conn(&app)?;
    let data = export::compose(&conn, &scope)?;
    let mut f = File::create(PathBuf::from(destination))?;
    crate::export::refi_qda::write_refi_qda(&data, &mut f)
}

#[tauri::command]
pub async fn export_stats(
    app: tauri::AppHandle,
    scope: ExportScope,
    format: StatsFormat,
    destination: String,
) -> AppResult<()> {
    let conn = project_conn(&app)?;
    let data = export::compose(&conn, &scope)?;
    let mut f = File::create(PathBuf::from(destination))?;
    match format {
        StatsFormat::Csv => crate::export::stats::write_stats_csv(&data, &mut f),
        StatsFormat::Markdown => crate::export::stats::write_stats_markdown(&data, &mut f),
    }
}

#[tauri::command]
pub async fn export_codebook(
    app: tauri::AppHandle,
    format: CodebookFormat,
    destination: String,
) -> AppResult<()> {
    let conn = project_conn(&app)?;
    let data = export::compose(&conn, &ExportScope::default())?;
    let mut f = File::create(PathBuf::from(destination))?;
    match format {
        CodebookFormat::Json => crate::export::codebook::write_codebook_json(&data, &mut f),
        CodebookFormat::Csv => crate::export::codebook::write_codebook_csv(&data, &mut f),
    }
}
